internal/validators/privileges/role_privileges.go

```diff
@@ -117,11 +109,91 @@ func getUserAndGroupPrincipals(ctx context.Context, username string, driver *vsp
 
 		header.Security, cerr = tokens.Issue(ctx, req)
 		if cerr != nil {
-			return "", nil, cerr
+			return nil, cerr
 		}
 	}
 
 	if err = ssoClient.Login(ssoClient.WithHeader(ctx, header)); err != nil {
+		return nil, err
+	}
+
+	return ssoClient, nil
+}
+
+func isAdminAccount(ctx context.Context, driver *vsphere.VSphereCloudDriver) (bool, error) {
+	ssoClient, err := configureSSOClient(ctx, driver)
+	if err != nil {
+		return false, err
+	}
+	defer ssoClient.Logout(ctx)
+
+	_, err = ssoClient.FindUser(ctx, driver.VCenterUsername)
+	if err != nil {
+		if strings.Contains(err.Error(), "NoPermission") {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func getPrivileges(ctx context.Context, driver *vsphere.VSphereCloudDriver, authManager *object.AuthorizationManager, username string) (map[string]bool, error) {
+	isAdmin, err := IsAdminAccount(ctx, driver)
+	if err != nil {
+		return nil, err
+	}
+
+	groupPrincipals := make([]string, 0)
```

Comment:
nit: initializing groupPrincipals down below in the non-admin case would be more clear, since it's overridden anyway in the admin case.

overall I think this function's readability would be improved if you added functions for privilege retrieval in both cases (e.g., getAdminPrivileges, getNonAdminPrivileges) and just returned one or the other based on the isAdmin var.

Reply:
I noticed some redundant if err checks and removed them to make this chunk of code a bit more clear.

IMO, reading down a function is more readable than jumping between a bunch of small functions.

Im hoping with the changes ive made it seems readable to you, but if not i can literally just make 2 small functions out of the things that are left in if isAdmin block and the stuff thats after that block.